keras_cv/models/segmentation/fcn.py

```diff
@@ -0,0 +1,355 @@
+# Copyright 2022 The KerasCV Authors
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import tensorflow as tf
+
+from keras_cv.models import VGG16
+from keras_cv.models import VGG19
+from keras_cv.layers.preprocessing import Resizing
+
+BACKBONE_CONFIG = {
+    "vgg16": {"BLOCK3": 11, "BLOCK4": 15, "BLOCK5": 19, "DENSE_START": -3},
+    "vgg19": {"BLOCK3": 12, "BLOCK4": 17, "BLOCK5": 22, "DENSE_START": -3},
+}
+
+
+class FCN(tf.keras.models.Model):
+    """A segmentation model based on the Fully Convolutional Network introduced by Long et. al.
+
+    Args:
+        classes: int, the number of classes for the segmentation model. Note that
+            the classes doesn't contain the background class, and the classes
+            from the data should be represented by integers with range
+            [0, classes).
+        backbone: a backbone network for the model. Can be a `tf.keras.models.Model`
+            instance. The supported pre-defined backbone models are:
+            1. "vgg16", a VGG16 model
+            2. "vgg19", a VGG19 model
+            Defaults to 'vgg16'.
+            Note: If a custom `tf.keras.models.Model` is passed, then only the
+            `Conv2D`, `MaxPooling2D` and `Dense` layers are extracted from it
+            to make the custom backbone.
+        model_architecture: str, defines the model architecture based on the
+            implementation details present in the paper. The supported
+            architectures are:
+            1. 'fcn8s', a FCN-8S definition
+            2. 'fcn16s', a FCN-16S definition
+            3. 'fcn32s', a FCN-32S definition
+            Defaults to 'fcn8s'.
+        include_resizing: bool, defines whether to include a `tf.keras.layers.Resizing` layer which defaults to a height and width of value 224. Defaults to `False`
+
+    """
+
+    def __init__(
+        self,
+        classes,
+        backbone="vgg16",
+        model_architecture=None,
+        include_resizing=False,
+        **kwargs
+    ):
+        super(FCN, self).__init__(**kwargs)
+
+        self.resize = include_resizing
+        self.num_classes = classes
+        self.backbone = backbone
+        self.model_architecture = model_architecture
+
+        if self.resize:
```

Comment:
image / segmentation map resizing should be moved to the data pipeline

Reply:
we have added resizing because vgg takes the default shape as 224x224x3

Reply:
Still, the resizing should be done outside of the model, and the backbone's input can be variable. A VGG can take any input size as long as it's not small enough (due to the MaxPooling calls), so the FCN should too.

Both FCN and VGG should be able to deal with varying input shapes